trajectory_executor/src/trajectory_executor/trajectory_executor.cpp

```diff
@@ -0,0 +1,188 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+#include "trajectory_executor/trajectory_executor.hpp"
+#include <ros/ros.h>
+#include <utility>
+#include <cav_msgs/SystemAlert.h>
+
+namespace trajectory_executor 
+{
+    cav_msgs::TrajectoryPlan trimFirstPoint(const cav_msgs::TrajectoryPlan plan) {
+        cav_msgs::TrajectoryPlan out(plan);
+        out.trajectory_points.erase(out.trajectory_points.begin());
+        return out;
+    }
+
+    TrajectoryExecutor::TrajectoryExecutor(int traj_frequency) :
+        _min_traj_publish_tickrate_hz(traj_frequency),
+        _timesteps_since_last_traj(0)
+    {
+    }
+
+    TrajectoryExecutor::TrajectoryExecutor() :
+        _min_traj_publish_tickrate_hz(10),
+        _timesteps_since_last_traj(0)
+    {
+    }
+
+    std::map<std::string, std::string> TrajectoryExecutor::queryControlPlugins()
+    {
+        // Hard coded stub for MVP since plugin manager won't be developed yet
+        // TODO: Query plugin manager to receive actual list of plugins and their corresponding topics
+        ROS_DEBUG("Executing stub behavior for plugin discovery MVP...");
+        std::map<std::string, std::string> out;
+
+        std::string default_control_plugin;
+        _default_nh->param<std::string>("default_control_plugin", default_control_plugin, "NULL");
+
+        std::string default_control_plugin_topic;
+        _default_nh->param<std::string>("default_control_plugin_topic", default_control_plugin_topic, "NULL");
+
+        out[default_control_plugin] = default_control_plugin_topic;
+        return out;
+    }
+    
+    void TrajectoryExecutor::onNewTrajectoryPlan(cav_msgs::TrajectoryPlan msg)
+    {
+        std::unique_lock<std::mutex> lock(_cur_traj_mutex); // Acquire lock until end of this function scope
+        ROS_DEBUG("Received new trajectory plan!");
+        ROS_DEBUG_STREAM("New Trajectory plan ID: " << msg.trajectory_id);
+        ROS_DEBUG_STREAM("New plan contains " << msg.trajectory_points.size() << " points");
+
+        _cur_traj = std::unique_ptr<cav_msgs::TrajectoryPlan>(new cav_msgs::TrajectoryPlan(msg));
+        ROS_DEBUG_STREAM("Successfully swapped trajectories!");
+
+        // TODO: Force traj emit
+        // TODO: Reset timer state
+    }
+
+    void TrajectoryExecutor::onTrajEmitTick(const ros::TimerEvent& te)
+    {
+        std::unique_lock<std::mutex> lock(_cur_traj_mutex);
+        ROS_DEBUG("TrajectoryExecutor tick start!");
+        
+
+        if (_cur_traj != nullptr) {
+            if (_timesteps_since_last_traj > 0) {
+                _cur_traj = std::unique_ptr<cav_msgs::TrajectoryPlan>(new cav_msgs::TrajectoryPlan(trimFirstPoint(*_cur_traj)));
+            }
+            if (!_cur_traj->trajectory_points.empty()) {
+                std::string control_plugin = _cur_traj->trajectory_points[0].controller_plugin_name;
```

Comment:
I think this assume each trajectory will only be sent to one control plugin. You might want to add a note to call it out.

Reply:
Yeah, at any given timestep the current trajectory is only sent to a single control plugin.